### ISSUE TITLE
[Bug] Fail decode namerecord string

### ISSUE THREAD

**Issue by moi15moi** [libass/libass#643]
Currently, libass always decode family name with utf-16be: https://github.com/libass/libass/blob/a48c98c274df25cc8d60cac1402948387d49840c/libass/ass_fontselect.c#L283-L284

But, Microsoft NameRecord don't always use utf-16be.
To know how libass should decode properly namerecord, see: https://github.com/MicrosoftDocs/typography-issues/issues/956#issuecomment-1205678068

Something like this could be added in ass_utils.c:
```c
char* get_name_encoding(FT_SfntName name) {
    if (name.platform_id == TT_PLATFORM_MICROSOFT)
    {
        switch (name.encoding_id)
        ​{
            case TT_MS_ID_PRC:
                return "windows-936";
            break;

            case TT_MS_ID_BIG_5:
                return (name.name_id == TT_NAME_ID_FONT_SUBFAMILY) ? "UTF-16BE" : "windows-950";
            break;

            case TT_MS_ID_WANSUNG:
                return (name.name_id == TT_NAME_ID_FONT_SUBFAMILY) ? "UTF-16BE" : "windows-949";
            break;

            default:
                return "UTF-16BE";
        }
    }
}
```

Finally, to decode byte into utf-8, libass could use ICU: https://unicode-org.github.io/icu/userguide/conversion/converters.html#1-single-string

PS: To test if it decode properly namerecord with BIG_5, download the font in this issue: https://github.com/MicrosoftDocs/typography-issues/issues/956#issue-1326452322

**Comment by astiob** [comment]
ICU is waaaay too massive for libass to use, if I’m not mixing anything up.

But the issue is, of course, real; thanks for creating a dedicated ticket for it. I already have name decoding code in https://github.com/astiob/libass/tree/debug-fonts. It should probably be adapted into mainline libass.

**Comment by moi15moi** [comment]
> ICU is waaaay too massive for libass to use, if I’m not mixing anything up.

Ok. I don't have a good knowledge of C.

> I already have name decoding code in https://github.com/astiob/libass/tree/debug-fonts.

I don't think it is a good idea to use mac platform id.

Here is what the Apple documentation mentions: [``Names with platformID 1 were required by earlier versions of macOS. Its use on modern platforms is discouraged``](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6name.html).

**Comment by astiob** [comment]
> Here is what the Apple documentation mentions: [`Names with platformID 1 were required by earlier versions of macOS. Its use on modern platforms is discouraged`](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM06/Chap6name.html).

That means about as much as the Microsoft docs not mentioning Windows 95 quirks that GDI nevertheless emulates to this day. It’s “discouraged” to use non-Unicode fonts at all, but that’s exactly what we’re trying to do here. (And IIRC macOS itself still preferred Macintosh-platform names when I last checked.)

That branch is (as the name suggests) meant for debugging font issues, so it dumps all the information it can find in a font. What we actually want (ideally) is [what VSFilter’s GDI calls use](https://github.com/libass/libass/wiki/Fonts-across-platforms#createfontindirectselectobject-vsfilter):

> For TrueType fonts, uses names with the same platform and encoding as the first valid Microsoft-platform cmap (if any) or MacRoman cmap (otherwise). Never uses Unicode-platform names.

so all Microsoft-platform encodings, as well as Macintosh-platform MacRoman.

**Comment by astiob** [comment]
Of course, we don’t currently support MacRoman cmaps, either, and I don’t remember if I’ve ever seen a font that lacked Microsoft-platform data and actually worked in VSFilter. I think Zapfino may have lacked them? Oughta retest it. But anyway, just Microsoft-platform names for now would be plenty good, to match our support of Microsoft-platform cmaps.

**Comment by moi15moi** [comment]
Here is 2 font that should not use utf-16be to be decoded.
[fonts.zip](https://github.com/libass/libass/files/11020004/fonts.zip)

**Comment by moi15moi** [comment]
I spoke with a Microsoft employee and he told me that GDI performed this processing:

```py
# This NameRecord is from 文鼎中特廣告體 - PlatEncID 4.ttf
name_record = NameRecord()
name_record.nameID = 1
name_record.string = b"\x00\xa4\x00\xe5\x00\xb9\x00\xa9\x00\xa4\x00\xa4\x00\xaf\x00S\x00\xbc\x00s\x00\xa7\x00i\x00\xc5\x00\xe9"
name_record.platformID = 3
name_record.platEncID = 4
name_record.langID = 0

encoding = get_name_record_encoding(name_record)

if name_record.platformID == 3 and encoding != "utf_16_be":
    name_to_decode = name_record.string.replace(b"\x00", b"")
else:
    name_to_decode = name_record.string

decoded_name = name_to_decode.decode(encoding)
```
